### DISCUSSION TITLE
Automatic Body Compression Clarification

### DISCUSSION THREAD

**Discussion by cinolt** [denoland/deno#14058] [Q&A] [answered]
It should be made clear in the manual how Deno internally handles automatic body compression:

1. Does Deno internally maintain a cache of compressed bodies?
2. If so, how long do those cache entries last?

If Deno does not maintain a cache, it could be inefficient for servers with high traffic to compress bodies upon every request, especially during a denial-of-service attack.

If Deno does maintain a cache, it is not specified how long cache entries last, and long-lived servers may end up keeping cache entries that are no longer needed.

A potential solution would be to specify that Deno does indeed utilize a cache, and to provide a Deno API to inspect/remove entries from such a cache.

**Comment by kitsonk** [comment, marked as answer]
It does not maintain a cache. It would a) be inefficient and b) break the model of the request being handled by the userland code.

The userland code would be significantly slower in a DDOS situation anyway, versus a per response compression.

If we were to implement response caching, it would align to the web platform APIs of `Cache` and `CacheStorage` and in those cases we could in theory sideload a compressed body in the response when a cached response was used. In those cases it would be up to the userland code to determine if the cached response is used or not anyway.

**Reply by cinolt** [reply]
I see; my only gripe is that compressing payloads every single time a resource is requested seems like a potentially colossal inefficiency, for some use cases.

For instance if a server is serving really big payloads (let's say 100MB+), it doesn't really make sense to compress that kind of payload a million times rather than one time.

But, this isn't the biggest deal because applications can always handle compression by itself, and automatic compression is still pretty handy for use cases where efficiency isn't of utmost importance.

**Reply by kitsonk** [reply]
> For instance if a server is serving really big payloads (let's say 100MB+), it doesn't really make sense to compress that kind of payload a million times rather than one time

Hopefully those wouldn't be sent as static bodies, but instead streamed, which currently is not compressed.

> But, this isn't the biggest deal because applications can always handle compression by itself, and automatic compression is still pretty handy for use cases where efficiency isn't of utmost importance.

"efficiency" is relative... serving huge static assets via a dynamic stack is inefficient, irrespective of them being compressed or not... trust me, if you are serving 100MB+ files with Deno, the compression in Rust wouldn't be the bottleneck.

**Reply by cinolt** [reply]
> "efficiency" is relative... serving huge static assets via a dynamic stack is inefficient, irrespective of them being compressed or not... trust me, if you are serving 100MB+ files with Deno, the compression in Rust wouldn't be the bottleneck.

Could you elaborate on what a "dynamic stack" is? My idea was to pre-allocate a 100MB+ Uint8Array then simply use that as a Response body each time. Seems like it would be pretty straightforward for Deno to send that as an HTTP response, so I don't know what's dynamic about that, unless Deno is doing something under the hood that would make that "inefficient"?
